library/core/class.format.php

```diff
@@ -2462,24 +2462,39 @@ public static function rich(string $deltas): string {
         return $renderer->render($blotGroups);
     }
 
+    const SAFE_PROTOCOLS = [
+        "http://",
+        "https://",
+        "tel:",
+        "mailto:",
+    ];
+
     /**
-     * Encode special CSS characters as hex.
+     * Sanitize a URL to ensure that it matches a whitelist of approved url schemes. If the url does not match one of these schemes, prepend `unsafe:` before it.
      *
-     * @param string $string
-     * @return mixed
+     * Allowed protocols
+     * - "http://",
+     * - "https://",
+     * - "tel:",
+     * - "mailto://",
+     *
+     * @param string $url The url to sanitize.
+     *
+     * @return string
      */
-    public static function cssSpecialChars(string $string) {
-        static $specialChars = [
-            "\\" => "\\00005c", "!" => "\\000021", "\"" => "\\000022", "#" => "\\000023", "$" => "\\000024",
-            "%" => "\\000025", "&" => "\\000026", "'" => "\\000027", "(" => "\\000028", ")" => "\\000029",
-            "*" => "\\00002a", "+" => "\\00002b", "," => "\\00002c", "-" => "\\00002d", "." => "\\00002e",
-            "/" => "\\00002f", ":" => "\\00003a", ";" => "\\00003b", "<" => "\\00003c", "=" => "\\00003d",
-            ">" => "\\00003e", "?" => "\\00003f", "@" => "\\000040", "[" => "\\00005b", "]" => "\\00005d",
-            "^" => "\\00005e", "`" => "\\000060", "{" => "\\00007b", "|" => "\\00007c", "}" => "\\00007d",
-            "~" => "\\00007e",
-        ];
+    public static function sanitizeUrl(string $url): string {
+        $isSafe = false;
 
-        $result = str_replace(array_keys($specialChars), array_values($specialChars), $string);
-        return $result;
+        foreach (self::SAFE_PROTOCOLS as $protocol) {
+            if (strpos($url, $protocol) === 0) {
```

Comment:
Would using [`parse_url`](http://php.net/parse_url) be a better fit here?

```
echo parse_url('http://google.com', PHP_URL_SCHEME) // "http"
echo parse_url('https://google.com', PHP_URL_SCHEME) // "https"
echo parse_url('mailto:user@example.com', PHP_URL_SCHEME) // "mailto"
echo parse_url('tel:5555555555', PHP_URL_SCHEME) // "tel"
```

You could grab the "scheme" and perform a simple `in_array` check. Beware the result of `parse_url` could be `null` or `false`.